flink/src/main/scala/ru/itclover/tsp/services/ArrowService.scala

```diff
@@ -0,0 +1,137 @@
+import org.apache.arrow.vector.ipc.{ArrowFileReader, ArrowReader, ArrowStreamReader, SeekableReadChannel}
+import org.apache.arrow.vector.{BaseValueVector, BitVector, Float4Vector, Float8Vector, IntVector, VarCharVector, BigIntVector, SmallIntVector}
+import org.apache.arrow.vector.types.Types
+import org.apache.arrow.vector.types.pojo.Schema
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+import scala.collection.mutable.ListBuffer
+
+
+object ArrowService {
+
+  /**
+  * Method for working with types from Apache Arrow schema.
+  */
+  def typesMap: Map[Types.MinorType, Class[_ >: Float with String with Boolean with Int with Double with Long with Short]] = Map(
+    Types.MinorType.BIGINT -> classOf[Long],
+    Types.MinorType.SMALLINT -> classOf[Short],
+    Types.MinorType.BIT -> classOf[Boolean],
+    Types.MinorType.INT -> classOf[Int],
+    Types.MinorType.VARCHAR -> classOf[String],
+    Types.MinorType.FLOAT4 -> classOf[Float],
+    Types.MinorType.FLOAT8 -> classOf[Double]
+  )
+
+  /**
+  * Method for retrieving schema and reader from input bytes
+    * @param inputData byte array with input data
+    * @return tuple with schema and reader
+    */
+  def retrieveSchemaAndReader(inputData: Array[Byte]): (Schema, ArrowReader, BaseAllocator) = {
+
+    val allocator = new RootAllocator(Integer.MAX_VALUE)
+    val bytesStream = new ByteArrayInputStream(inputData)
+
+    val reader = new ArrowStreamReader(bytesStream, allocator)
+
+    (reader.getVectorSchemaRoot.getSchema, reader, allocator)
+
+  }
+
+  /**
+  * Method for retrieving schema and reader from input file
+    * @param inputData file with input data
+    * @return tuple with schema and reader
+    */
+  def retrieveSchemaAndReader(inputData: File): (Schema, ArrowReader, BaseAllocator) = {
+
+    val allocator = new RootAllocator(Integer.MAX_VALUE)
+    val fileStream = new FileInputStream(inputData)
+    val readChannel = new SeekableReadChannel(fileStream.getChannel)
+
+    val reader = new ArrowFileReader(readChannel, allocator)
+
+    (reader.getVectorSchemaRoot.getSchema, reader, allocator)
+
+  }
+
+  /**
+  * Method for converting input to list of values
+    * @param input arrow schema, arrow reader, arrow allocator
+    * @return list of values
+    */
+  def convertData(input: (Schema, ArrowReader, BaseAllocator)): ListBuffer[ListBuffer[(String, Any)]] = {
+
+    val schema = input._1
+    val reader = input._2
+    val allocator = input._3
+
+    val schemaFields: List[String] = schema.getFields
+                                           .asScala
+                                           .map(_.getName)
+                                           .toList
+
+    val schemaRoot = reader.getVectorSchemaRoot
+    var rowCount = 0
+
+    var readCondition = reader.loadNextBatch()
+
+    var result: mutable.ListBuffer[mutable.ListBuffer[(String, Any)]] = mutable.ListBuffer.empty
+
+    while(readCondition){
```

Comment:
wouldn't it be called only once?

Reply:
no, it will be updated in while loop. Line 126